sys-utils/ipcutils.c

```diff
@@ -477,6 +644,105 @@ void ipc_msg_free_info(struct msg_data *msgds)
 	}
 }
 
+int posix_ipc_msg_get_info(const char *name, struct posix_msg_data **msgds)
+{
+	FILE *f;
+	DIR *d;
+	int i = 0;
+	struct posix_msg_data *p;
+	struct dirent *de;
+
+	p = *msgds = xcalloc(1, sizeof(struct msg_data));
+	p->next = NULL;
+
+	if (access(_PATH_DEV_MQUEUE, F_OK) != 0) {
+		warnx(_("%s does not seem to be mounted. Use 'mount -t mqueue none %s' to mount it."), _PATH_DEV_MQUEUE, _PATH_DEV_MQUEUE);
+		return -1;
+	}
+
+	d = opendir(_PATH_DEV_MQUEUE);
+	if (!d) {
+		warnx(_("cannot open %s"), _PATH_DEV_MQUEUE);
+		return -1;
+	}
+
+	while ((de = readdir(d)) != NULL) {
+		if ((de->d_name[0] == '.' && de->d_name[1] == '\0') ||
+		    (de->d_name[0] == '.' && de->d_name[1] == '.' &&
+		     de->d_name[2] == '\0'))
+			continue;
+
+		struct stat st;
+		char path[PATH_MAX];
+		snprintf(path, sizeof(path), "%s/%s", _PATH_DEV_MQUEUE, de->d_name);
+
```

Comment:
And here :-)

Reply:
The mentioned changes were made in `posix_ipc_shm_get_info` and `posix_ipc_sem_get_info`.

In `posix_ipc_msg_get_info`:
It was noted that for SHMs and semaphores the object name `/test_name` and `test_name`  (also `//test_name`) pertain to the same structure. While mqueues need to strictly follow mq_overview(8) convention of starting the object name with a /, otherwise throwing an EINVAL.

So I assume getting rid of the `char* path` wouldn't be applicable in here. Furthermore, I we would need the file path to read the QSIZE parameter from the file (to be stored in (struct posix_msg_data *p)->c_bytes).
I highlight few changes I made to it.

```suggestion
struct stat st;
char path[PATH_MAX];
snprintf(path, sizeof(path), "%s/%s", _PATH_DEV_MQUEUE, de->d_name);

if (fstatat(dirfd(d), de->d_name, &st, 0) < 0) {
	continue;
}

f = fopen(path, "r");
if (!f) {
	continue;
}
if (fscanf(f, "QSIZE:%"SCNu64, &p->q_cbytes) != 1) {
	fclose(f);
	continue;
}
fclose(f);

memset(path, 0, sizeof(path));
path[0] = '/';
strcat(path, de->d_name);

p->mname = xstrdup(path);
```

Reply:
Ah, I had no clue about '/' convention. Thanks,

However, you can still avoid using path[] in your code by using fstatat() and  openat(), and fdopen() + fscanf(). This will make your code more robust and readable. Frankly, the combination of opendir() and snprintf() is a bad coding pattern and will stand out to any reviewer :)  

The *at() function was introduced for this pattern; it avoids extra path buffer, kernel path lookup, and possible races. If you already have the directory, use it :-)
                                                                               
If '/' is required, then use `p->name = strconcat("/", de->d_name);` (see lib/strutils.c)